Make sure to not restart application-muted microphone tracks in case the audio unit gets unsuspended

youennf opened the pull request:
#### 439677029dc8350cb7a128fa96cc51f8b0c59496
<pre>
Make sure to not restart application-muted microphone tracks in case the audio unit gets unsuspended
<a href="https://bugs.webkit.org/show_bug.cgi?id=243248">https://bugs.webkit.org/show_bug.cgi?id=243248</a>
rdar://97612254

BaseAudioSharedUnit unmuted all sources currently.
This overrides sources that were muted by the application.

To prevent this, at the time the unit gets interrupted, we mute the sources but store whether these sources were producing data.
Then, when unsuspending, we only unmute sources that were producing data.

To be able to write a test, we move the interruption code from iOS to all platforms.
We add IPC to allow Internals to trigger interruptions in GPUProcess through IPC.

Covered by newly added test LayoutTests/fast/mediastream/keep-microphone-muted-on-uninterruption.html.

Reviewed by Eric Carlson.

* LayoutTests/fast/mediastream/keep-microphone-muted-on-uninterruption-expected.txt: Added.
* LayoutTests/fast/mediastream/keep-microphone-muted-on-uninterruption.html: Added.
* Source/WebCore/platform/audio/AudioSession.cpp:
(WebCore::AudioSession::beginInterruption):
(WebCore::AudioSession::endInterruption):
* Source/WebCore/platform/audio/AudioSession.h:
* Source/WebCore/platform/mediastream/ios/CoreAudioCaptureSourceIOS.h:
* Source/WebCore/platform/mediastream/mac/BaseAudioSharedUnit.cpp:
(WebCore::BaseAudioSharedUnit::addClient):
(WebCore::BaseAudioSharedUnit::removeClient):
(WebCore::BaseAudioSharedUnit::clearClients):
(WebCore::BaseAudioSharedUnit::resume):
(WebCore::BaseAudioSharedUnit::suspend):
* Source/WebCore/platform/mediastream/mac/CoreAudioCaptureSource.cpp:
(WebCore::CoreAudioCaptureSource::create):
(WebCore::CoreAudioCaptureSource::createForTesting):
(WebCore::CoreAudioCaptureSourceFactory::CoreAudioCaptureSourceFactory):
(WebCore::CoreAudioCaptureSourceFactory::~CoreAudioCaptureSourceFactory):
(WebCore::CoreAudioCaptureSourceFactory::unit):
(WebCore::CoreAudioCaptureSourceFactory::beginInterruption):
(WebCore::CoreAudioCaptureSourceFactory::endInterruption):
(WebCore::CoreAudioCaptureSource::startProducingData):
* Source/WebCore/platform/mediastream/mac/CoreAudioCaptureSource.h:
(WebCore::CoreAudioCaptureSource::canResumeAfterInterruption const):
(WebCore::CoreAudioCaptureSource::setCanResumeAfterInterruption):
(WebCore::CoreAudioCaptureSourceFactory::setOverrideUnit):
* Source/WebCore/testing/Internals.cpp:
(WebCore::Internals::beginAudioSessionInterruption):
(WebCore::Internals::endAudioSessionInterruption):
* Source/WebCore/testing/Internals.h:
* Source/WebCore/testing/Internals.idl:
* Source/WebKit/GPUProcess/media/RemoteAudioSessionProxy.cpp:
(WebKit::RemoteAudioSessionProxy::triggerBeginInterruptionForTesting):
(WebKit::RemoteAudioSessionProxy::triggerEndInterruptionForTesting):
* Source/WebKit/GPUProcess/media/RemoteAudioSessionProxy.h:
* Source/WebKit/GPUProcess/media/RemoteAudioSessionProxy.messages.in:
* Source/WebKit/WebProcess/GPU/media/RemoteAudioSession.cpp:
(WebKit::RemoteAudioSession::beginInterruptionForTesting):
(WebKit::RemoteAudioSession::endInterruptionForTesting):
* Source/WebKit/WebProcess/GPU/media/RemoteAudioSession.h:

Canonical link: <a href="https://commits.webkit.org/252905@main">https://commits.webkit.org/252905@main</a>
</pre>

eric-carlson commented on Source/WebCore/platform/audio/AudioSession.cpp:
```diff
@@ -124,6 +124,8 @@ void AudioSession::removeInterruptionObserver(InterruptionObserver& observer)
 
 void AudioSession::beginInterruption()
 {
+    RELEASE_LOG_ERROR(WebRTC, "AudioSession::beginInterruption %u", m_interruptionObservers.computeSize());
```
Should this really be logged as an error?

youennf replied:
This should be removed, I did it in the new patch

eric-carlson commented on Source/WebCore/platform/audio/AudioSession.cpp:
```diff
@@ -135,6 +137,8 @@ void AudioSession::beginInterruption()
 
 void AudioSession::endInterruption(MayResume mayResume)
 {
+    RELEASE_LOG_ERROR(WebRTC, "AudioSession::endInterruption");
```
Ditto

youennf replied:
Ditto

youennf replied:
This should be removed, I did it in the new patch

eric-carlson commented on Source/WebCore/platform/mediastream/mac/BaseAudioSharedUnit.cpp:
```diff
@@ -44,6 +44,8 @@ BaseAudioSharedUnit::BaseAudioSharedUnit()
 
 void BaseAudioSharedUnit::addClient(CoreAudioCaptureSource& client)
 {
+    RELEASE_LOG_ERROR(WebRTC, "BaseAudioSharedUnit::addClient %p", this);
```
Ditto

youennf replied:
This should be removed, I did it in the new patch

eric-carlson commented on Source/WebCore/platform/mediastream/mac/BaseAudioSharedUnit.cpp:
```diff
@@ -52,6 +54,8 @@ void BaseAudioSharedUnit::addClient(CoreAudioCaptureSource& client)
 
 void BaseAudioSharedUnit::removeClient(CoreAudioCaptureSource& client)
 {
+    RELEASE_LOG_ERROR(WebRTC, "BaseAudioSharedUnit::removeClient %p", this);
```
Ditto

youennf replied:
This should be removed, I did it in the new patch

eric-carlson commented on Source/WebCore/platform/mediastream/mac/BaseAudioSharedUnit.cpp:
```diff
@@ -60,6 +64,8 @@ void BaseAudioSharedUnit::removeClient(CoreAudioCaptureSource& client)
 
 void BaseAudioSharedUnit::clearClients()
 {
+    RELEASE_LOG_ERROR(WebRTC, "BaseAudioSharedUnit::clearClients %p", this);
```
Ditto

youennf replied:
This should be removed, I did it in the new patch

achristensen07 committed:
Revert "Remove legacy build system script phases and build settings"
https://bugs.webkit.org/show_bug.cgi?id=243221

Unreviewed.

It broke the iOS build.

* Source/JavaScriptCore/Configurations/Base.xcconfig:
* Source/JavaScriptCore/Configurations/JavaScriptCore.xcconfig:
* Source/JavaScriptCore/Configurations/ToolExecutable.xcconfig:
* Source/JavaScriptCore/JavaScriptCore.xcodeproj/project.pbxproj:
* Source/ThirdParty/ANGLE/ANGLE.xcodeproj/project.pbxproj:
* Source/ThirdParty/ANGLE/Configurations/ANGLE-dynamic.xcconfig:
* Source/ThirdParty/ANGLE/Configurations/Base.xcconfig:
* Source/ThirdParty/libwebrtc/Configurations/Base.xcconfig:
* Source/ThirdParty/libwebrtc/Configurations/libabsl.xcconfig:
* Source/ThirdParty/libwebrtc/Configurations/libwebrtc.xcconfig:
* Source/ThirdParty/libwebrtc/libwebrtc.xcodeproj/project.pbxproj:
* Source/WTF/Configurations/Base.xcconfig:
* Source/WTF/Configurations/WTF.xcconfig:
* Source/WTF/WTF.xcodeproj/project.pbxproj:
* Source/WebCore/PAL/Configurations/Base.xcconfig:
* Source/WebCore/PAL/Configurations/PAL.xcconfig:
* Source/WebCore/PAL/PAL.xcodeproj/project.pbxproj:
* Source/WebGPU/Configurations/Base.xcconfig:
* Source/WebGPU/Configurations/WGSL.xcconfig:
* Source/WebKit/Configurations/Base.xcconfig:
* Source/WebKit/Configurations/WebKit.xcconfig:
* Source/WebKit/WebKit.xcodeproj/project.pbxproj:
* Source/WebKitLegacy/WebKitLegacy.xcodeproj/project.pbxproj:
* Source/WebKitLegacy/mac/Configurations/Base.xcconfig:
* Source/WebKitLegacy/mac/Configurations/WebKitLegacy.xcconfig:
* Tools/DumpRenderTree/mac/Configurations/Base.xcconfig:
* Tools/DumpRenderTree/mac/Configurations/DumpRenderTreeLibrary.xcconfig:
* Tools/DumpRenderTree/mac/Configurations/LayoutTestHelper.xcconfig:
* Tools/TestWebKitAPI/Configurations/Base.xcconfig:
* Tools/TestWebKitAPI/Configurations/TestWTFLibrary.xcconfig:

Canonical link: https://commits.webkit.org/252875@main

emw-apple committed:
Remove legacy build system script phases and build settings
https://bugs.webkit.org/show_bug.cgi?id=243221
rdar://89701972

Reviewed by Alexey Proskuryakov.

Delete WK_WHICH_BUILD_SYSTEM and WK_USE_NEW_BUILD_SYSTEM settings.
Remove script phases which checked these settings and do nothing under
XCBuild.

* Source/JavaScriptCore/Configurations/Base.xcconfig:
* Source/JavaScriptCore/Configurations/JavaScriptCore.xcconfig:
* Source/JavaScriptCore/Configurations/ToolExecutable.xcconfig:
* Source/JavaScriptCore/JavaScriptCore.xcodeproj/project.pbxproj:
* Source/ThirdParty/ANGLE/ANGLE.xcodeproj/project.pbxproj:
* Source/ThirdParty/ANGLE/Configurations/ANGLE-dynamic.xcconfig:
* Source/ThirdParty/ANGLE/Configurations/Base.xcconfig:
* Source/ThirdParty/libwebrtc/Configurations/Base.xcconfig:
* Source/ThirdParty/libwebrtc/Configurations/libabsl.xcconfig:
* Source/ThirdParty/libwebrtc/Configurations/libwebrtc.xcconfig:
* Source/ThirdParty/libwebrtc/libwebrtc.xcodeproj/project.pbxproj:
* Source/WTF/Configurations/Base.xcconfig:
* Source/WTF/Configurations/WTF.xcconfig:
* Source/WTF/WTF.xcodeproj/project.pbxproj:
* Source/WebCore/PAL/Configurations/Base.xcconfig:
* Source/WebCore/PAL/Configurations/PAL.xcconfig:
* Source/WebCore/PAL/PAL.xcodeproj/project.pbxproj:
* Source/WebGPU/Configurations/Base.xcconfig:
* Source/WebGPU/Configurations/WGSL.xcconfig:
* Source/WebKit/Configurations/Base.xcconfig:
* Source/WebKit/Configurations/WebKit.xcconfig:
* Source/WebKit/WebKit.xcodeproj/project.pbxproj:
* Source/WebKitLegacy/WebKitLegacy.xcodeproj/project.pbxproj:
* Source/WebKitLegacy/mac/Configurations/Base.xcconfig:
* Source/WebKitLegacy/mac/Configurations/WebKitLegacy.xcconfig:
* Tools/DumpRenderTree/mac/Configurations/Base.xcconfig:
* Tools/DumpRenderTree/mac/Configurations/DumpRenderTreeLibrary.xcconfig:
* Tools/DumpRenderTree/mac/Configurations/LayoutTestHelper.xcconfig:
* Tools/TestWebKitAPI/Configurations/Base.xcconfig:
* Tools/TestWebKitAPI/Configurations/TestWTFLibrary.xcconfig:

Canonical link: https://commits.webkit.org/252870@main

youennf committed:
Make sure to not restart application-muted microphone tracks in case the audio unit gets unsuspended
https://bugs.webkit.org/show_bug.cgi?id=243248
rdar://97612254

BaseAudioSharedUnit unmuted all sources currently.
This overrides sources that were muted by the application.

To prevent this, at the time the unit gets interrupted, we mute the sources but store whether these sources were producing data.
Then, when unsuspending, we only unmute sources that were producing data.

To be able to write a test, we move the interruption code from iOS to all platforms.
We add IPC to allow Internals to trigger interruptions in GPUProcess through IPC.

Covered by newly added test LayoutTests/fast/mediastream/keep-microphone-muted-on-uninterruption.html.

Reviewed by Eric Carlson.

* LayoutTests/fast/mediastream/keep-microphone-muted-on-uninterruption-expected.txt: Added.
* LayoutTests/fast/mediastream/keep-microphone-muted-on-uninterruption.html: Added.
* Source/WebCore/platform/audio/AudioSession.cpp:
(WebCore::AudioSession::beginInterruption):
(WebCore::AudioSession::endInterruption):
* Source/WebCore/platform/audio/AudioSession.h:
* Source/WebCore/platform/mediastream/ios/CoreAudioCaptureSourceIOS.h:
* Source/WebCore/platform/mediastream/mac/BaseAudioSharedUnit.cpp:
(WebCore::BaseAudioSharedUnit::addClient):
(WebCore::BaseAudioSharedUnit::removeClient):
(WebCore::BaseAudioSharedUnit::clearClients):
(WebCore::BaseAudioSharedUnit::resume):
(WebCore::BaseAudioSharedUnit::suspend):
* Source/WebCore/platform/mediastream/mac/CoreAudioCaptureSource.cpp:
(WebCore::CoreAudioCaptureSource::create):
(WebCore::CoreAudioCaptureSource::createForTesting):
(WebCore::CoreAudioCaptureSourceFactory::CoreAudioCaptureSourceFactory):
(WebCore::CoreAudioCaptureSourceFactory::~CoreAudioCaptureSourceFactory):
(WebCore::CoreAudioCaptureSourceFactory::unit):
(WebCore::CoreAudioCaptureSourceFactory::beginInterruption):
(WebCore::CoreAudioCaptureSourceFactory::endInterruption):
(WebCore::CoreAudioCaptureSource::startProducingData):
* Source/WebCore/platform/mediastream/mac/CoreAudioCaptureSource.h:
(WebCore::CoreAudioCaptureSource::canResumeAfterInterruption const):
(WebCore::CoreAudioCaptureSource::setCanResumeAfterInterruption):
(WebCore::CoreAudioCaptureSourceFactory::setOverrideUnit):
* Source/WebCore/testing/Internals.cpp:
(WebCore::Internals::beginAudioSessionInterruption):
(WebCore::Internals::endAudioSessionInterruption):
* Source/WebCore/testing/Internals.h:
* Source/WebCore/testing/Internals.idl:
* Source/WebKit/GPUProcess/media/RemoteAudioSessionProxy.cpp:
(WebKit::RemoteAudioSessionProxy::triggerBeginInterruptionForTesting):
(WebKit::RemoteAudioSessionProxy::triggerEndInterruptionForTesting):
* Source/WebKit/GPUProcess/media/RemoteAudioSessionProxy.h:
* Source/WebKit/GPUProcess/media/RemoteAudioSessionProxy.messages.in:
* Source/WebKit/WebProcess/GPU/media/RemoteAudioSession.cpp:
(WebKit::RemoteAudioSession::beginInterruptionForTesting):
(WebKit::RemoteAudioSession::endInterruptionForTesting):
* Source/WebKit/WebProcess/GPU/media/RemoteAudioSession.h:

Canonical link: https://commits.webkit.org/252905@main
